From 7a63ed17a9ee8970972715a10e227c41e22ff341 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:09:08 -0400
Subject: [usb] cdc-acm: fix runtime PM imbalance at shutdown

Message-id: <1410887463-20674-477-git-send-email-dzickus@redhat.com>
Patchwork-id: 95460
O-Subject: [RHEL7 PATCH 476/591] USB: cdc-acm: fix runtime PM imbalance at shutdown
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 5292afa657d0e790b7479ad8eef9450c1e040b3d
Author: Johan Hovold <jhovold@gmail.com>
Date:   Mon May 26 19:23:45 2014 +0200

    USB: cdc-acm: fix runtime PM imbalance at shutdown

    Make sure only to decrement the PM counters if they were actually
    incremented.

    Note that the USB PM counter, but not necessarily the driver core PM
    counter, is reset when the interface is unbound.

    Fixes: 11ea859d64b6 ("USB: additional power savings for cdc-acm devices
    that support remote wakeup")

    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/class/cdc-acm.c b/drivers/usb/class/cdc-acm.c
index eddeba6..6bbd203 100644
--- a/drivers/usb/class/cdc-acm.c
+++ b/drivers/usb/class/cdc-acm.c
@@ -590,12 +590,13 @@ static void acm_port_shutdown(struct tty_port *port)
  struct urb *urb;
  struct acm_wb *wb;
  int i;
+ int pm_err;
 
  dev_dbg(&acm->control->dev, "%s\n", __func__);
 
  mutex_lock(&acm->mutex);
  if (!acm->disconnected) {
-  usb_autopm_get_interface(acm->control);
+  pm_err = usb_autopm_get_interface(acm->control);
   acm_set_control(acm, acm->ctrlout = 0);
 
   for (;;) {
@@ -613,7 +614,8 @@ static void acm_port_shutdown(struct tty_port *port)
   for (i = 0; i < acm->rx_buflimit; i++)
    usb_kill_urb(acm->read_urbs[i]);
   acm->control->needs_remote_wakeup = 0;
-  usb_autopm_put_interface(acm->control);
+  if (!pm_err)
+   usb_autopm_put_interface(acm->control);
  }
  mutex_unlock(&acm->mutex);
 }
-- 
1.7.1